python/nav/web/auth/ldap.py

```diff
@@ -121,11 +122,66 @@ def open_ldap():
     return lconn
 
 
-def authenticate(login, password):
+def authenticate(username, password):
     """
-    Attempt to authenticate the login name with password against the
-    configured LDAP server.  If the user is authenticated, required
-    group memberships are also verified.
+    Authenticate the username and password against the configured LDAP server.
+
+    Required group memberships are also verified.
+
+    Returns an authenticated Account with updated groups, or None.
+    """
+    if not available:
+        return None
+    ldap_user = get_ldap_user(username, password)
+    try:
+        account = Account.objects.get(login__iexact=username, ext_sync='ldap')
+    except Account.DoesNotExist:
+        if ldap_user:
+            account = autocreate_ldap_user(ldap_user, password)
+            return account
+    if account.locked:
+        _logger.info("Locked user %s tried to log in", account.login)
+        return None
+    if account.check_password(password):
+        account = update_ldap_user(ldap_user, account, password)
+        return account
```

Comment:
`ldap_user` can be False, and then `update_ldap_user` will raise an exception. It is unclear what the best solution is though:

Change the `if` to:

```
if ldap_user and account.check_password(password):
```

If ldap_user is False authenticate will return None.

Split the `if` into two:

```
if ldap_user:
    account = update_ldap_user(ldap_user, account, password)

if account.check_password(password):
    return account
```

Worse, the latter `if` might not be needed at all because if ldap_user is True, we have already been authenticated with ldap, so authenticating again is unneccessary.

A third way is:

```
if not ldap_user:
    return None

account = update_ldap_user(ldap_user, account, password)
```

and then again choose whether to check the password again...